pkg/kubeadm/node/discovery.go

```diff
@@ -0,0 +1,65 @@
+package kubenode
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/square/go-jose"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
+	kubeadmapi "k8s.io/kubernetes/pkg/kubeadm/api"
+)
+
+func RetrieveTrustedClusterInfo(params *kubeadmapi.BootstrapParams) (*clientcmdapi.Config, error) {
+	apiServerURL, err := url.Parse(strings.Split(params.Discovery.ApiServerURLs, ",")[0])
+	if err != nil {
+		return nil, err
+	}
+
+	host, port := strings.Split(apiServerURL.Host, ":")[0], 9898
+
+	req, err := http.NewRequest("GET", fmt.Sprintf("http://%s:%d/cluster-info/v1/?token-id=%s", host, port, params.Discovery.TokenID), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	buf := new(bytes.Buffer)
+	io.Copy(buf, res.Body)
+	res.Body.Close()
+
+	object, err := jose.ParseSigned(buf.String())
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := object.Verify(params.Discovery.Token)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterInfo := kubeadmapi.ClusterInfo{}
+
+	if err := json.Unmarshal(output, &clusterInfo); err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("ClusterInfo: %#v", clusterInfo)
+
+	if len(clusterInfo.CertificateAuthorities) == 0 || len(clusterInfo.Endpoints) == 0 {
+		return nil, fmt.Errorf("Cluster discovery object (ClusterInfo) is invalid")
+	}
+
+	// TODO figure out what we should do when there is a chain of certificates and more then one API endpoint
+	apiServer := clusterInfo.Endpoints[0]
+	caCert := []byte(clusterInfo.CertificateAuthorities[0])
```

Comment:
We shouldn't get back a chain here but rather a set of roots.  This enables root certificate rotation.  We need to configure the client to validate the server if it is signed by any of the returned certificates.